doc/compiler/optimizer/InliningAndInliningGuards.md

```diff
@@ -0,0 +1,406 @@
+n11616n         ==>aloadi
+n12449n     aconst 0xd5a390 (methodPointerConstant sharedMemory )
+```
+
+If there is only one extended class `B`, or if the profiling information shows the receiver type is 100% `B`,
+it might be safe to use a VFT test or a class pointer comparison guard. If more than one subclass is loaded
+such as B, D, E, etc., and none of them overrides `B.foo()`, this is a sign to use a method test rather than
+a VFT test. If the top value is 35% which is below the threshold of 75%, but its subclasses show 25% of `D`
+and 25% of `E`, and three of them add up to more than 75% and all three classes map to `B.foo()`, the inliner
+will choose the method comparison guard because it will succeed 75% of the time, but the class comparison
+guard will have less chance of succeeding.
+
+## Nonoverridden Guard (Explicit Guard or Nopable guard)
+
+Nonoverridden Guard can either be an explicit guard or a nopable guard.
+There is [a bit in every J9Method](https://github.com/eclipse-openj9/openj9/blob/8393a7277cbd664acc8c9477d66b2dc9a9d6f412/runtime/compiler/env/j9method.cpp#L5074-L5077)
+indicating whether the method is overridden or not.  Any time a subclass is loaded, the VM checks if the method
+in that class overrides the method in the super class and it sets that bit for the method.
+
+```
+n11271n   iflcmpne --> block_34 BBStart at n11265n (inlineNonoverriddenGuard )
+n11270n     land
+n11267n       lload      0xc28068[#1492  Static] [flags 0x307 0x0 ]
+n11268n       lconst 4
+n11269n     lconst 0
+```
+
+The method can be inlined with an explicit guard that checks the bit of the method in IL trees. If it is not
+overridden, the inlined code is executed, otherwise it invokes the virtual call.
+
+The same test can be nopable. When it gets to code gen (`virtualGuardHelper()`), it checks if the method is overridden
+or not at compile time. If not, test code is not generated, and instead a runtime assumption is generated.
+The runtime assumption can be interpreted as "if this method ever gets overridden, go patch this location to go
+to the slow path”. It stores the information of where it is in the code and which method it depends on. Before
+a class is being loaded, the JIT runs through all the relevant runtime assumptions and patches the locations.
+
+Virtual guard noping can be disabled by setting `-Xjit:disableVirtualGuardNOPing`. There are also cases where
+we do not want to use nopable guard. During start-up, all kinds of classes are still being loaded and the class
+hierarchy has not settled down. By using nopable guard too early on, we would pay the price for patching because
+after the guard is patched, the execution goes to the slow path forever. In this case, we might choose to be more
+conservative and use explicit guards although they are less efficient. At least some proportion of the time,
+the execution could go to the inlined path.
+
+## Single Abstract Implementation Guard (Nopable guard)
```

Comment:
Should this section be merged somehow with a prior section that talked about abstract and interface calls ?